### PR TITLE
Reduce calls to list stores in replica allocation

### PR DESCRIPTION
**Pull request by kimchy**
If all nodes that we want to allocate the replica on will end up being throttled,
then there is no need to try and go and list the stored for it, as it will not be allocated
anyhow. Note, this does not change the amount of calls needed in total, since we cache the
result of each list store, but, it will reduce the number of list stores calls we do in each
allocation round.

### CODE REVIEW

File: src/main/java/org/elasticsearch/gateway/GatewayAllocator.java [hunk 1]
```diff
@@ -264,8 +264,14 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
         while (unassignedIterator.hasNext()) {
             MutableShardRouting shard = unassignedIterator.next();
 
+            // if its not a primary, no need to check for this shard at all, this should only handle replicas
+            if (shard.primary()) {
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: +1

File: src/main/java/org/elasticsearch/gateway/GatewayAllocator.java [hunk 2]
```diff
@@ -274,13 +280,17 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
                 // if we can't allocate it on a node, ignore it, for example, this handles
                 // cases for only allocating a replica after a primary
                 Decision decision = allocation.deciders().canAllocate(shard, node, allocation);
+                if (decision.type() != Decision.Type.THROTTLE) {
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: I'm a bit confused by this. If `canBeAllocatedToAtLeastOneNode` is false,  we already skip this shard. If `canBeAllocatedToAtLeastOneNode` is true, it means that at least one decision was YES, which also implies not THROTTLE, meaning that allAreThrottled will also be false. This makes allAreThrottled redundant? or am I missing something?

File: src/main/java/org/elasticsearch/gateway/GatewayAllocator.java (hunk 1 above)
Comment (s1monw) [inline src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: this one sounds familiar :)

File: src/main/java/org/elasticsearch/gateway/GatewayAllocator.java (hunk 2 above)
Comment (kimchy) [inline src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: actually, yea, I think you are right, it makes this optimization invalid, argh, I already thought about it before :)

**Pull request closed without merging**